### PR TITLE
Added Resolver for Solana and Ethereum Domain Name Websites[REVISED]

### PR DESCRIPTION
**Pull request by mjkid221**
# Solana and Ethereum Name Service Resolver for Backpack.
<hr/>
EDIT:
<b>
This is a revised version of the pull request at #2259. I ran into some issues with the older build in #2259 so I started a fresh PR.
<b/>
<hr/>

Solana and Ethereum Domain Name Services (DNS) provide a decentralized and cost-effective way to link domain names to corresponding on-chain data, such as websites associated with those domain names.

This pull request adds support for resolving Solana (SNS) and Ethereum (ENS) domain names in the URL, allowing users to automatically redirect to their website stored in the domain content. The feature build add support for IPFS, IPNS, and regular HTTPS websites (and even IP addesses!), making Backpack a more versatile and desirable crypto wallet. This implementation was greatly inspired from `zhvng`'s [sns-resolver](https://github.com/zhvng/sns-resolver) which was deprecated. 

The new feature integrates with the existing wallet functionality, and utilizes the Solana (Bonfida) and Ethereum (Ethers) library to lookup domain name content and redirect to attached websites upon validation. All invalid URLs will redirect to an error 404 page. Users can also configure the gateway to use of their own choice (including custom url) to resolve their domain website if desired so.

This is probably the first time someone has ever created a domain name website resolver for Solana at the wallet level, and I believe it will greatly enhance the user experience and make Backpack a more powerful tool for the crypto community.

Please review the code and let me know if there are any issues or suggestions for improvements. I would be happy to make any necessary changes to ensure this feature gets merged into the master branch. Thank you!

## Process to Use
1. Set content for SNS or ENS with URL of your desired website. 
    - You can resolve to either IPFS OR IPNS via setting the field to either one of `ipfs=`, `ipns=`, `ipfs://`, `ipns://` respectively. 
    - You can also resolve to a regular `https://` or `http://` website by just setting the field with a regular url if the underlying domain name service supports it. 

2. Assuming Backpack is already installed, go to your URL tab and search for your `<domainName>.sol` or `<domainName>.eth`.
   - It will automatically redirect as well if the search domain is prefixed with `http://` or `https://` such as `http://<domainName>.eth` or postfixed with `/` such as `<domainName>.sol/`.
   - Redirects also work with both SOL and ETH subdomains but it is important to note that:
       -  ENS allows for multiple nested child subdomains which is resolvable by the implementation. Technically, it should be able to resolve infinite(not really) amount of child subdomains. 
       - Bonfida's SNS does NOT allow nested subdomains and so you cannot have a subdomain that owns a subdomain, thus cannot resolve to it. 

## Search Engines
`zhvng` had implemented a way to redirect the browser from searching domain names in the search engine in the address bar since `.sol` or `.eth` is not a recognised domain extension. This new implementation adds support for the most popular search engines `Google`, `DuckDuckGo` and additionally, `Bing`.

### Redirecting SOL or ETH Domain
https://user-images.githubusercontent.com/103424726/214815673-957f7567-9f90-4283-ab7c-7cfb921e0113.mov

### Choosing Different Gateways from the `Preferences` page of Backpack and Redirecting to a ETH Subdomain
https://user-images.githubusercontent.com/103424726/214817943-849b82f0-cad9-43b7-b26c-266e98b7fc40.mov

### CODE REVIEW

File: packages/common/src/preferences.ts [hunk 1]
```diff
@@ -1,3 +1,5 @@
+import { DEFAULT_GATEWAY } from "@coral-xyz/app-extension/src/redirects/constants";
```
Comment (armaniferrante) [inline packages/common/src/preferences.ts:1]: This is a circular import. We should define DEFAULT_GATEWAY inside common and import it from app-extension.

Comment (mjLabrys) [reply packages/common/src/preferences.ts:1]: Hi Armani, thanks for the feedback. I'll take a look at resolving the issue once I'm finished with work. Let me know if there's anything else that needs changing. I also saw your tweet about the feature and I'm glad it's received a positive response. If you don't mind, I wrote a thread on Twitter recently about the importance of on-chain DNS resolvers, including a mention of Backpack and this feature. Check it out here: https://twitter.com/mjkid221/status/1618587248722857984. Please do excuse my shameless shilling haha...

Comment (mjLabrys) [reply packages/common/src/preferences.ts]: The circular dependency issue has been resolved now :)

**Comment by mjLabrys** [comment]
Merged latest development branch into this working branch again, and resolved merge conflicts.

**Comment by armaniferrante** [comment]
> Merged latest development branch into this working branch again, and resolved merge conflicts.

Thank you!

**Comment by mjLabrys** [comment]
@hkirat @armaniferrante I was wondering if I could please receive some guidance or feedback regarding the next steps involved in preparing my pull request for merging. I have resolved merge conflicts again and would appreciate any advice or direction you can offer.

**Comment by mjLabrys** [comment]
@dr497, thanks for the feedback. Yeah, I've been busy with other works lately and haven't had the opportunity to update this code base for a while. Will see if I can get to it again asap

**Comment by mjLabrys** [comment]
Just wanted to provide an update. Now Backpack redirection can be toggled per supported network like the following. Tested this works with Brave browser as intended. This should give users flexibility to use their preferred resolution method.
Screenshot 2023-05-01 at 9 56 06 pm

**Comment by mjLabrys** [comment]
I've updated the PR with a new implementation that resolves URL, IPFS, Shadow drive, and Arweave records in this priority set by Bonfida's Solana domain records. This adheres to the standard established by Bonfida. For reference, please refer https://github.com/Bonfida/name-resolver to understand how this works.

In addition to the priority resolution of URL, IPFS, Shadow drive, and Arweave records, the implementation can still resolve to any form of URL set in the domain content. However, this resolution will only occur after checking the content of the URL, IPFS, Shadow drive, and Arweave records.

Directly Referencing Bonfida's Documentation:
To make your domain browsable, you must set your records:
1. Go to your domain's page (e.g., https://naming.bonfida.org/domain/bonfida)
2. Connect your wallet
    * In order for your domain to resolve, you must set one of the following records:
    * IPFS: Use the format ipfs://{CID} (e.g., ipfs://QmZk9uh2mqmXJFKu2Hq7kFRh93pA8GDpSZ6ReNqubfRKKQ)
    * Arweave: Use the format arw://{HASH} (e.g., arwv://KuB5jmew87_M2flH9f6ZpB9jlDv8hZSHPrmGUY8KqEk)
    * URL: Use the format url_to_your_website (e.g., https://bonfida.org)
    * ShadowDrive: Use the format shdw://{HASH}

**Pull request closed without merging**